### ISSUE TITLE
Add the possibility to specify per-file compilation database path

### ISSUE THREAD

**Issue by matemat13** [clangd/coc-clangd#182]
Hello, I'm investigating switching from YouCompleteMe (YCM) to coc.vim and so far I like what I see. However, I cannot get coc.vim working with our C++ project structure, which is unfortunately a stopper. The problem is that in our case the `compile_commands.json` file for a `*.cpp` source are not located above the source file in the folder hierarchy, but in a `build` subfolder of an above folder. E.g.:

- source file: `<workspace_path>/src/<package_name>/src/<filename>.cpp`
- compilation flags file: `<workspace_path>/build/<package_name>/compile_commands.json`

```
<workspace_path>
├── build
│   └── <package_name>
│       └── compile_commands.json
└── src
    └── <package_name>
        └── src
            └── <filename>.cpp
```

In YCM we use a custom global `.ycm_extra_conf.py` script to point `clang`/`clangd` to the correct `compile_commands.json` file based on the source file's name and path. I don't see a way to do a similar thing in coc.vim/coc-clangd. The closest way to replicate this behavior AFAIK is to use a custom `.vim/coc-settings.json` for each package and setup the paths manually, which is simply unfeasable for us and would pollute our code repositories with IDE-specific garbage.

**TL;DR:** Is there a way to point `clangd` to the correct `compile_commands.json` file based on the source file's name and path when it cannot be found automatically (even if it meant forking this repo and modifying it)? If not, could it be added?

**Comment by kadircet** [comment]
Yes there is, (thanks to @sam-mccall)!

Before diving into that I would like to mention that this exact case you mentioned (i.e. compile_commands.json is under a folder called precisely `build` in an ancestor path) should already be handled by clangd, starting with v11. In times prior to that, the solution was usually creating a symlink in one of the ancestors paths of the source files to the compile_comands.json placed in an arbitrary directory.

For a more complex scenario, we've recently (it means you'll need a clangd built from head) introduced a config support for compilation database discovery. You can put a configuration file at `$XDG_CONFIG_HOME/clangd/config.yaml` with contents:
```
If:
  PathMatch: /path/to/project1/.*
CompileFlags:
  CompilationDatabase: /path/to/project1s_compilation_database_dir/
---
If:
  PathMatch: /path/to/project2/.*
CompileFlags:
  CompilationDatabase: /path/to/project2s_compilation_database_dir/
```
Note that this configuration is not dynamic, so you'll need to know absolute path for the compile_commands.json's directory in advance. If you can afford a per-project clangd configuration (which can be checked in to the repository), you can have relative path support, and get rid of the `If` block by putting a `.clangd` file at the root of your source tree with contents:
```
CompileFlags:
  CompilationDatabase: compilation_database_dir/  # Note that this is relative to `.clangd` file.
```

You can find more details about clangd configuration support at https://clangd.llvm.org/config.html. Unfortunately it doesn't contain information about the `CompilationDatabase` yet.

**Comment by matemat13** [comment]
> Yes there is, (thanks to @sam-mccall)!
> 
> Before diving into that I would like to mention that this exact case you mentioned (i.e. compile_commands.json is under a folder called precisely `build` in an ancestor path) should already be handled by clangd, starting with v11.

Aaaah that's so close and would be such an elegant solution! Unfortunately in our case, the `compile_commands.json` file is in a subfolder of the `build` folder (I messed up the folder tree in the original post - updated now). Maybe I'll try somehow automatically merging the different jsons into one and putting it in the `build` folder directly after each build.

> For a more complex scenario, we've recently (it means you'll need a clangd built from head) introduced a config support for compilation database discovery. You can put a configuration file at `$XDG_CONFIG_HOME/clangd/config.yaml` (...)

I'll look into that, hopefully it'll be flexible enough for our purposes. Thanks a lot!

**Comment by fannheyward** [comment]
You can use `clangd.compilationDatabaseCandidates` to special the dirs that contains compile_commands.json.
